Sources/AsyncHTTPClient/ConnectionPool/State Machine/StateMachine.swift

```diff
@@ -0,0 +1,275 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+extension HTTPConnectionPool {
+    struct StateMachine {
+        struct Action {
+            let request: RequestAction
+            let connection: ConnectionAction
+
+            init(_ request: RequestAction, _ connection: ConnectionAction) {
+                self.request = request
+                self.connection = connection
+            }
+
+            static var none: Action {
+                return .init(.none, .none)
+            }
```

Comment:
What's the difference between this and just allowing returned actions to be `nil`?

Reply:
In my opinion, making this an optional make everything harder. The option to switch like I do on an enum was also added more recently. Before you would need to do this:

```swift
enum Action {
  case foo
  case bar
}

func run(_ action: Action?) {
  switch action {
  case .some(.foo):
    break
  case .some(.bar):
    break
  case .none:
    break
  }
}
```

Which is ugly. Also autocomplete works much nicer, if it is not an optional, but an enum with a `none` case.